proposals/more-const-types.md

```diff
@@ -0,0 +1,78 @@
+
+It could be expanded to include any `readonly struct` type.
+
+We should consider if a `ref struct` type, such as `ReadOnlySpan<T>` could be a compile-time constant.
+
+## Motivation
+[motivation]: #motivation
+
+Only constant expressions can be used in pattern matching expressions. This expands the types that can be used in patterns. For example, it's common to use GUIDs as identifiers. This feature enables matching on GUID values. Currently, the only way to do that is to convert a GUID to a ReadOnlySpan<byte> and use a list pattern on the list of bytes.
+
+Similarly, any other type that holds multiple values, like record struct types, could be used to match patterns.  For example, a `Point` type could be matched agains a `const` for the origin.
+
+Other uses for constant expressions include:
+
+- default argument values.
+- Attribute parameters
+
+## Detailed design
+[design]: #detailed-design
+
+This is the bulk of the proposal. Explain the design in enough detail for somebody familiar with the language to understand, and for somebody familiar with the compiler to implement,  and include examples of how the feature is used. This section can start out light before the prototyping phase but should get into specifics and corner-cases as the feature is iteratively designed and implemented.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+## Alternatives
+[alternatives]: #alternatives
+
+What other designs have been considered? What is the impact of not doing this?
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What parts of the design are still undecided?
+
+- **What types should be allowed?**:  Preventing mutation for types whose members are all unmanaged types (or `struct` types whose members are all unmanaged types) is a manageable problem: a `const` instance can't be accessed as a writable `ref`. However, if any member is a reference type, that problem is harder to solve.
+- **Are there implications for the runtime?** Depending on how constants are stored in memory, is there an impact on the runtime?
+- **What are the implications of allowing constant ref structs? Are they even useful?**: Are the rules surrounding ref safety consistent with where constants are used?
+- **How will down-level scenarios work?** Can these constants be accessed by code using previous compilers? What about using Reflection to access fields?
+- **Does a type need to "opt in" to all constant definitions?** (Otherwise, changes later could introduce a field that disallows all existing `const` declarations):
+
+   ```csharp
+    public struct Point
+    {
+        public int X;
+        public int Y;
+    }
+
+    // V2:
+    
+    public struct Point
+    {
+        public int X;
+        public int Y;
+        public object SomeReference;
+    }
+    ```
+
```

Comment:
How do we reconcile this with the runtime concept of field constant?  

Currently C# constants match the runtime's version of field constants.  For example, a field can be declared that is a constant which is different than a static field with a value. The constant value is encoded in the metadata which the compiler uses to emit as load instructions at the use site instead of referencing the field at runtime, because no field exists for it at runtime.  The kinds of values that can be field constants is constrained to those that can be encoded in metadata. I'm not sure what those are, but it probably does not include arbitrary types even if they are blittable.

Of course, C# could deviate from the runtime and allow other types to be constant but would have to change to use static fields for these instead of field constants.

Reply:
C# already diverges from the runtime's concept of constants. Consider `decimal`:

```cs
const decimal d = 1.0M;
```

Results in:

```
    // Fields
    .field public static initonly valuetype [System.Runtime]System.Decimal d
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.DecimalConstantAttribute::.ctor(uint8, uint8, uint32, uint32, uint32) = (
        01 00 01 00 00 00 00 00 00 00 00 00 0a 00 00 00
        00 00
    )
```